infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java

```diff
@@ -49,24 +45,44 @@ public boolean isServerReadyToHandleMsg(CorfuMsg msg) {
      * @param r   The router that took in the message.
      */
     public void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        if (shutdown) {
+        if (getState() == ServerState.SHUTDOWN) {
             log.warn("Server received {} but is already shutdown.", msg.getMsgType().toString());
             return;
         }
 
+        if (!isServerReadyToHandleMsg(msg)) {
+            r.sendResponse(ctx, msg, CorfuMsgType.NOT_READY.msg());
+            return;
+        }
+
         if (!getHandler().handle(msg, ctx, r)) {
             log.warn("Received unhandled message type {}", msg.getMsgType());
         }
     }
 
+    protected void setState(ServerState state){
+        this.state.set(state);
+    }
+
+    public ServerState getState(){
+        return state.get();
+    }
+
     public abstract ExecutorService getExecutor();
 
     /**
      * Shutdown the server.
      */
     public void shutdown() {
-        shutdown = true;
+        setState(ServerState.SHUTDOWN);
         getExecutor().shutdownNow();
     }
 
+    /**
+     * The server state.
+     * Represents server in a particular state: READY, NOT_READY, SHUTDOWN etc.
+     */
+    public enum ServerState {
+        READY, NOT_READY, SHUTDOWN
```

Comment:
Just curious when is NOT_READY state used since I don't see it being set?

Reply:
It's not used yet, it will be used at least in LogUnitServer::resetLogUnit() (next pr-s)

Reply:
ok thanks